nginx resource: audit the nginx binary and how it was compiled

rx294 opened the pull request:
Signed-off-by: Rony Xavier <rx294@nyu.edu>
Signed-off-by: Aaron Lippold <lippold@gmail.com>

aaronlippold committed:
nginx base resource and the nginx_module resource.

Signed-off-by: Aaron Lippold <lippold@gmail.com>
Signed-off-by: Rony Xavier <rx294@gmail.com>

adamleff commented on docs/resources/nginx.md.erb:
```diff
@@ -0,0 +1,74 @@
+
+## Supported Properties
+
+* `'prefix'`, `'service'`, `'user'`, `'group'`, `'version'`, `'conf_path'`, `'conf_dir'`,
```
Please alphabetize these with each property being a separate bullet.

adamleff commented on docs/resources/nginx.md.erb:
```diff
@@ -0,0 +1,74 @@
+`openssl_version ` returns a hash with 'version' and 'date' as keys
+
+    describe nginx.openssl_version['date'] do
+      it { should eq '11 Feb 2013' }
```
please move the `openssl_version` to an `its`:

```
describe nginx do
  its('openssl_version.date') { should eq '11 Feb 2013' }
end
```

rx294 replied:
Hey Adam, thank you for the review. I appreciate your pointers.
 As for the use of 'openssl_version.date' the, 'openssl_version' returns a hash but I am unable to reach the date data without using ' nginx.openssl_version['date']'

I really wanted to use the dot syntax, please let me know if you can show my how to implement that feature.

Thank you

rx294 replied:
Please disregard...I found your solution below. Thank you again!

adamleff commented on docs/resources/nginx.md.erb:
```diff
@@ -0,0 +1,74 @@
+
+`compiler_info ` returns a hash with 'compiler' , version' and 'date' as keys
+
+    describe nginx.compiler_info['compiler'] do
```
Same as above...
```
its('compiler_info.compiler')
```

adamleff commented on docs/resources/nginx.md.erb:
```diff
@@ -0,0 +1,74 @@
+`pid_file ` returns an inspec file object for the nginx pid file
+
+    describe nginx.pid_file do
+       its('owner'){ should eq 'root' }
```
```
its('pid_file.owner')
```

adamleff commented on docs/resources/nginx_module.md.erb:
```diff
@@ -0,0 +1,39 @@
+`loaded_modules ` returns an array of loaded modules for the nginx instance
+
+    describe nginx_module.loaded_modules do
+      it { should include 'http_ssl' }
```
```
its('loaded_modules') { should include 'http_ssl' }
```

adamleff commented on lib/resources/nginx.rb:
```diff
@@ -0,0 +1,108 @@
+    "
+
+    def initialize
+      return skip_resource 'The `nginx` resource is not yet tested on your OS.' if inspec.os.windows?
```
"not yet available" vs. "not yet tested"

adamleff commented on lib/resources/nginx.rb:
```diff
@@ -0,0 +1,108 @@
+      end
+    "
+
+    def initialize
```
Should probably accept an optional parameter of where the `nginx` binary can be found as it may not be in the InSpec user's path.

adamleff commented on lib/resources/nginx.rb:
```diff
@@ -0,0 +1,108 @@
+    # @todo RFC: which of these methods should be private and or not exposed?
+    #
+
+    def method_missing(name)
```
Since there is a known number of methods/keys we offer, I'd prefer that we define each of those as methods rather than using `method_missing`

adamleff commented on lib/resources/nginx.rb:
```diff
@@ -0,0 +1,108 @@
+      @params[name.to_s] unless @params.nil?
+    end
+
+    def read_content
```
If there's a chance this could get called more than once for a given resource, we need to avoid reparsing the data each time.

Something like:

```ruby
return unless @params.empty?
parse_config
...
```

adamleff commented on lib/resources/nginx.rb:
```diff
@@ -0,0 +1,108 @@
+    def parse_config
+      @params['prefix'] = @data.scan(/--prefix=(\S+)\s/).flatten.first
+      @params['service'] = 'nginx'
+      @params['user'] = @data.scan(/--user=(\S+)\s/).flatten.first
```
This may be duplicative data with the planned nginx_config resource in #1889. Given that we're running `nginx -V` without accepting a config path, we should probably remove values that are settable in the `nginx.conf` from this resource and let the new nginx_config resource take over.

aaronlippold replied:
No, if #1889 parses out these values, we think the `nginx_conf` resource should just get them from the base `nginx` resource just like `apache` and others do.

aaronlippold replied:
Also, the base resource is looking at the running process, which is what `nginx -V` reports on. I think the assumption that we would use the `nginx_conf` resource to look at values in the conf file and we use the `nginx` resource to look at the active settings of the nginx process is a good point of view.

adamleff replied:
Unfortunately, you are mistaken regarding `nginx -V`. It does not show anything about the running `nginx` process but rather how it was configured during compile time as well as version information. The `user` parameter here is referring to the default username to run nginx, as compiled into the binary. If I run nginx with a different user, `nginx -V` will **not** reflect that.

rx294 replied:
Thank for the note Adam, we have removed all the duplicate data from the nginx resource. 
Also, please note that nginx_module resource has been separated from this PR and has been submitted as a different PR. 

Thank you for all the reviews, please take a look at the corrections when you get a chance.

adamleff commented on lib/resources/nginx.rb:
```diff
@@ -0,0 +1,108 @@
+
+    def openssl_version
+      result = @data.scan(/built with OpenSSL\s(\S+)\s(\d\d\s\S+\s\d\d\d\d)/).flatten
+      openssl_version = { 'version' => result[0], 'date' => result[1] }
```
Create this as a `Hashie::Mash` rather than a straight hash so someone can do `its('openssl_version.version')`

adamleff commented on lib/resources/nginx.rb:
```diff
@@ -0,0 +1,108 @@
+
+    def compiler_info
+      result = @data.scan(/built by (\S+)\s(\S+)\s(\S+)/).flatten
+      compiler_info = { 'compiler' => result[0], 'version' => result[1], 'date' => result[2] }
```
Create this as a `Hashie::Mash` rather than a straight hash so someone can do `its('compiler_info.version')`

adamleff commented on lib/resources/nginx.rb:
```diff
@@ -0,0 +1,108 @@
+    def compiler_info
+      result = @data.scan(/built by (\S+)\s(\S+)\s(\S+)/).flatten
+      compiler_info = { 'compiler' => result[0], 'version' => result[1], 'date' => result[2] }
+      compiler_info unless compiler_info.nil?
```
`compiler_info` can never be nil... you're always creating a hash, though some values may be nil.

Change this to:

```
    def compiler_info
      result = @data.scan(/built by (\S+)\s(\S+)\s(\S+)/).flatten
      Hashie::Mash.new({ 'compiler' => result[0], 'version' => result[1], 'date' => result[2] })
    end
```

adamleff commented on lib/resources/nginx.rb:
```diff
@@ -0,0 +1,108 @@
+
+    def support_info
+      support_info = @data.scan(/(.*\S+) support enabled/).flatten.join(' ')
+      support_info unless support_info.nil?
```
This will never be nil, and you want to return nil if it's not available.

I'd write this to be:

```ruby
support_info = @data.scan(/(.*\S+) support enabled/).flatten
support_info.empty? ? nil : support_info.join(' ')
```

adamleff commented on lib/resources/nginx.rb:
```diff
@@ -0,0 +1,108 @@
+
+    def pid_file
+      pid_file = inspec.file(@params['pid_path'])
+      pid_file unless pid_file.nil?
```
How could this be nil?

adamleff commented on lib/resources/nginx.rb:
```diff
@@ -0,0 +1,108 @@
+      'Nginx Environment'
+    end
+
+    private :read_content, :parse_config, :parse_path, :parse_http_path
```
Our standard is to move the methods below a `private` declaration so it's clearer. Could you please make that change?

adamleff commented on lib/resources/nginx_module.rb:
```diff
@@ -0,0 +1,35 @@
+
+    def initialize(modulename = nil)
+      @module = modulename
+      return skip_resource 'The `nginx_module` resource is not yet tested on your OS.' if !inspec.os.linux?
```
"not yet available" vs. "not yet tested"

adamleff commented on lib/resources/nginx_module.rb:
```diff
@@ -0,0 +1,35 @@
+      "
+
+    def initialize(modulename = nil)
+      @module = modulename
```
Like your `nginx` resource, you probably need to take an optional argument for the full path to the `nginx` binary.

adamleff commented on lib/resources/nginx_module.rb:
```diff
@@ -0,0 +1,35 @@
+    def loaded_modules
+      cmd = inspec.command('nginx -V 2>&1')
+      modules_list = cmd.stdout.to_s.scan(/--with-(\S+)_module/).flatten
+      cmd.exit_status.zero? ? modules_list : nil
```
If the command returns non-zero, you'll return a nil, which will throw an error in `loaded?` since there's no `include?` method on a NilClass object.

I'd return an empty array here instead of `nil`

adamleff commented on test/unit/resources/nginx_test.rb:
```diff
@@ -0,0 +1,124 @@
+      resource = load_resource('nginx')
+      _(resource.pid_path).must_match '/var/run/nginx.pid'
+    end
+
```
Remove empty lines between `end`s please.

adamleff reviewed (changes requested):
Hi again @rx294 - thanks for your PR. I've left you feedback to address. Thank you!

adamleff commented on lib/resources/nginx.rb:
```diff
@@ -0,0 +1,108 @@
+# copyright: 2017
+# author: Aaron Lippold, lippold@gmail.com
+# author: Rony Xavier, rx294@gmail.com
+# license: All rights reserved
```
Please remove this line. InSpec is released under Apache 2.0. We have a pending PR to remove this line from all our files (acknowledging that you probably copy/pasted this from another existing file 🙂 ).

adamleff commented on lib/resources/nginx_module.rb:
```diff
@@ -0,0 +1,35 @@
+# encoding: utf-8
+# author: Rony Xavier, rx294@gmail.com
+# author: Aaron Lippold, lippold@gmail.com
+# license: All rights reserved
```
Please remove this line. InSpec is released under Apache 2.0. We have a pending PR to remove this line from all our files (acknowledging that you probably copy/pasted this from another existing file 🙂 ).

aaronlippold replied:
done

rx294 committed:
Seperated Nginx and Nginx_module resources into different PRs
All requested changes made from the PR review

Signed-off-by: Rony Xavier <rx294@nyu.edu>

rx294 committed:
Merge branch 'master' into al/nginx_resource

rx294 committed:
PR for Nginx resource

Signed-off-by: Rony Xavier <rx294@nyu.edu>

aaronlippold committed:
updated the resource to be a bit more verbose on finding core data elements.
fixed a bit of documntation formatting.

Signed-off-by: Aaron Lippold <lippold@gmail.com>

aaronlippold committed:
Merge branch 'master' into al/nginx_resource

Signed-off-by: Aaron Lippold <lippold@gmail.com>

aaronlippold committed:
fixed rubocop errors and condenced the class size down.

Signed-off-by: Aaron Lippold <lippold@gmail.com>

aaronlippold committed:
fixed unit tests and mock commands / files
added proper input and error checking
guarded param passing and added validation

Signed-off-by: Aaron Lippold <lippold@gmail.com>

aaronlippold committed:
removed copyright lines

Signed-off-by: Aaron Lippold <lippold@gmail.com>

aaronlippold committed:
removed unneed mock cmd file

Signed-off-by: Aaron Lippold <lippold@gmail.com>

aaronlippold committed:
Merge branch 'master' into al/nginx_resource

aaronlippold committed:
fixed a conflict in helper.rb

Signed-off-by: Aaron Lippold <lippold@gmail.com>

aaronlippold committed:
fixed deleted line for oracle

Signed-off-by: Aaron Lippold <lippold@gmail.com>

adamleff commented on lib/resources/nginx.rb:
```diff
@@ -0,0 +1,125 @@
+    attr_reader :params, :bin_dir
+
+    def initialize(nginx_bin_dir = nil)
+      return skip_resource 'The `nginx` resource is not yet available on your OS.' if inspec.os.windows?
```
Per our review today, we're going to simplify the discovery of the nginx binary to user-supplied OR look in path (i.e. use `inspec.command('nginx').exist?`). If we don't have a usable binary, we'll `skip_resource` to give the user feedback that we can't use.

adamleff commented on lib/resources/nginx.rb:
```diff
@@ -0,0 +1,125 @@
+
+    private
+
+    def bin_dir_from_command
```
This method can likely go away once we simplify the `initialize` method

adamleff commented on lib/resources/nginx.rb:
```diff
@@ -0,0 +1,125 @@
+      inspec.command('which nginx | cut -d/ -f1,2,3').stdout.strip
+    end
+
+    def nginx_conf_from_command
```
Per our discussion, we're going to remove conf_path and conf_dir from the params, so this method can likely go away as well. I appreciate what you were trying to do though!

adamleff commented on lib/resources/nginx.rb:
```diff
@@ -0,0 +1,125 @@
+      cmd.exit_status.zero? ? cmd.stdout.chomp : nil
+    end
+
+    def locate_by_path(type = nil)
```
Once we simplify the `initialize` method, this method can likely go away, or at least be slimmed down.

adamleff commented on test/unit/mock/cmd/bash-c-type-nginx:
```diff
@@ -0,0 +1 @@
+nginx is /usr/sbin/nginx
```
Please check if this is necessary after we slim down the `initialize` method

rx294 replied:
this is needed by the command('nginx').exists? call

adamleff replied:
This could be mocked in the tests though, no?

rx294 replied:
Hey adam I am bit confused...are suggesting a test for the  command('nginx').exists?

adamleff replied:
Not quite... I'm suggesting that you can mock the `inspec.command` call with a minitest mock so that the InSpec MockLoader stuff never gets called. Then you have full control over the responses returned to the method when you're testing.

You can look at the unit tests for the `host` resource for some examples.

 * set up an `inspec` object: https://github.com/chef/inspec/blob/master/test/unit/resources/host_test.rb#L190
 * stub it for the test: https://github.com/chef/inspec/blob/master/test/unit/resources/host_test.rb#L194-L196
 * stub the "command(blah).exist?" calls: https://github.com/chef/inspec/blob/master/test/unit/resources/host_test.rb#L194-L196

adamleff replied:
This reduces the need for those `type` command mocks, and lets you write test for both when the command exists and when it doesn't.

rx294 replied:
@adamleff I have resolved the merge conflicts. I am having issue with the mock. Perhaps I can implement it in the future resources if that's ok with you. Thank you!

adamleff commented on test/unit/mock/cmd/ps-aux-nginx-awk:
```diff
@@ -0,0 +1 @@
+/etc/nginx/nginx.conf
```
This mock is no longer necessary once the conf_path parameter is removed.

adamleff commented on test/unit/mock/cmd/which-nginx-cut-d-123:
```diff
@@ -0,0 +1 @@
+/usr/sbin
```
This mock is no longer necessary once we slim down the auto-discovery of the nginx binary.

adamleff commented on test/unit/resources/nginx_test.rb:
```diff
@@ -0,0 +1,101 @@
+      resource = load_resource('nginx')
+      _(resource.service).must_match 'nginx'
+    end
+    it 'Verify nginx parsing `conf_path` - `/etc/nginx`' do
```
This test and the one beneath it will likely need to be removed - please check after your refactor! 🙂

adamleff reviewed (changes requested):
This is really close, thanks!

rx294 committed:
Updates as per remarks
Merged Nginx & Nginx_modules

Signed-off-by: Rony Xavier <rx294@nyu.edu>

rx294 committed:
Merge branch 'master' into al/nginx_resource

Signed-off-by: Rony Xavier <rx294@nyu.edu>

adamleff commented on lib/resources/nginx.rb:
```diff
@@ -13,20 +13,22 @@ class Nginx < Inspec.resource(1)
-        return skip_resource 'Error fining the NGINX binary or `bin_dir`, plese review the provided `PATH` is only the `base_dir` and or ensure the `nginx` binary is in the system `PATH`.'
-      end
-      cmd = inspec.command("#{bin_dir}/nginx -V 2>&1")
+      path ||= '/usr/sbin/'
```
Let's be consistent with other resources... rather than taking a directory and then joining it to `nginx`, the user should supply a full path to the nginx binary.

```ruby
def initialize(nginx_path = '/usr/sbin/nginx')
  return skip_resource 'blah blah..."
end
```

rx294 replied:
Done. Thank you @adamleff.

adamleff reviewed (changes requested):
The changes you implemented look good, but I'd like to require the user to supply the full path to the `nginx` binary rather than just a directory where it exists.

rx294 committed:
Updates as per reccomendations

Signed-off-by: Rony Xavier <rx294@nyu.edu>

aaronlippold committed:
Merge branch 'master' of https://github.com/chef/inspec into al/nginx_resource

aaronlippold committed:
Merge branch 'master' of https://github.com/chef/inspec into al/nginx_resource

Signed-off-by: Rony Xavier <rx294@nyu.edu>

rx294 committed:
Merge branch 'al/nginx_resource' of https://github.com/aaronlippold/inspec into al/nginx_resource

Signed-off-by: Rony Xavier <rx294@nyu.edu>

rx294 committed:
Merge branch 'al/nginx_resource' of https://github.com/aaronlippold/inspec into al/nginx_resource

rx294 committed:
Merge branch 'al/nginx_resource' of https://github.com/aaronlippold/inspec into al/nginx_resource

Signed-off-by: Rony Xavier <rx294@nyu.edu>

adamleff commented on test/helper.rb:
```diff
@@ -360,6 +360,14 @@ def md.directory?
+      %{/usr/sbin/nginx -V 2>&1} => cmd.call('nginx-v'),
+      %{bash -c 'type "/usr/sbin/nginx"'} => cmd.call('bash-c-type-nginx'),
+      # needed for two differnt inspec.command call formats
+      # host resource: dig commands
```
Lines 368-370 are not related to this PR and are probably from a bad merge conflict resolution given the duration this PR has been open. Can you please remove them?

adamleff commented on docs/resources/nginx.md.erb:
```diff
@@ -0,0 +1,72 @@
+An `nginx` InSpec audit resource block extracts configuration settings that should be tested:
+
+    describe nginx do
+      its('attribute'){ should eq 'value' }
```
Space missing between `)`and `{`

adamleff reviewed (changes requested):
@rx294 two quick clean-up items and then this one should be good to approve.

rx294 committed:
Updates

Signed-off-by: Rony Xavier <rx294@nyu.edu>

chris-rock reviewed (approved):
👍 Thank you @rx294 and @aaronlippold for this new resource. Great review @adamleff 
LGTM
